test/qa_data_sink.cpp

```diff
@@ -0,0 +1,688 @@
+    T                  next_value         = {};
+    std::size_t        next_tag           = 0;
+    std::vector<tag_t> tags; // must be sorted by index, only one tag per sample
+
+    void
+    init(const property_map &, const property_map &) {
+        // optional init function that is called after construction and whenever settings change
+        fair::graph::publish_tag(out, { { "n_samples_max", n_samples_max } }, n_tag_offset);
+    }
+
+    constexpr std::int64_t
+    available_samples(const Source &) noexcept {
+        const auto ret = static_cast<std::int64_t>(n_samples_max - n_samples_produced);
+        // forcing one sample, at a time, see below
+        return ret > 0 ? 1 : -1; // '-1' -> DONE, produced enough samples
+    }
+
+    T
+    process_one() noexcept {
+        if (next_tag < tags.size() && tags[next_tag].index <= static_cast<std::make_signed_t<std::size_t>>(n_samples_produced)) {
+            tag_t &out_tag = this->output_tags()[0];
+            // TODO when not enforcing single samples in available_samples, one would have to do:
+            // const auto base = std::max(out.streamWriter().position() + 1, tag_t::signed_index_type{0});
+            // out_tag        = tag_t{ tags[next_tag].index - base, tags[next_tag].map };
+            // Still think there could be nicer API to set a tag from process_one()
+            out_tag = tag_t{ 0, tags[next_tag].map };
+            this->forward_tags();
+            next_tag++;
+        }
+
+        n_samples_produced++;
+        return next_value++;
+    }
+};
+
+/**
+ * Example tag observer (TriggerObserver implementation) for the multiplexed listener case (interleaved data). As a toy example, we use
+ * data tagged as Year/Month/Day.
+ *
+ * For each of year, month, day, the user can specify whether:
+ *
+ *  - option not set: The field is to be ignored
+ *  - -1: Whenever a change between the previous and the current tag is observed, start a new data set (StopAndStart)
+ *  - other values >= 0: A new dataset is started when the tag matches, and stopped, when a tag doesn't match
+ *
+ * (Note that the TriggerObserver is stateful and remembers the last tag seen, other than a stateless TriggerPredicate)
+ */
+struct Observer {
+    std::optional<int>                       year;
+    std::optional<int>                       month;
+    std::optional<int>                       day;
+    std::optional<std::tuple<int, int, int>> last_seen;
+    bool                                     last_matched = false;
+
+    explicit Observer(std::optional<int> year_, std::optional<int> month_, std::optional<int> day_) : year(year_), month(month_), day(day_) {}
+
+    static inline bool
+    same(int x, std::optional<int> other) {
+        return other && x == *other;
+    }
+
+    static inline bool
+    changed(int x, std::optional<int> other) {
+        return !same(x, other);
+    }
+
+    trigger_observer_state
+    operator()(const tag_t &tag) {
+        const auto ty = tag.get("Y");
```

Comment:
For readability sake:
  * 'Y' -> `Year`
  * `M' -> `Month`
  * `D` -> `Day`
  
  This helps new users/readers to get into the gist of things.